### PR TITLE
EES-3695 Add GetContact and UpdateContact endpoints

### PR DESCRIPTION
**Pull request by markhiveit**
This PR adds new endpoints to get and update contacts on a publication, and also does other various bits of cleanup in preparation for switching to the new Admin dashboard:

- Adds a GetContact endpoint. This endpoint has an optional query parameter so services can request permissions if required.
- Adds a UpdateContact endpoint.
- Make the Publications table's `ContactId` nonnullable. Since a publication should always have a contact, there isn't any reason for the column to be nullable. :rotating_light: :rotating_light: :rotating_light: Due to this change, we need to double check there aren't any instances of `ContactId` that are null on Prod before deploying :rotating_light: :rotating_light: :rotating_light: I already updated all entries with null on Dev, and Test, Preprod and Prod didn't have any instances of null.
- Create `PublicationCreateRequest` `PublicationCreateViewModel`. This is necessary as updating the contact is no longer possible via the update publication endpoint, but when creating a publication you must also create an associated contact, as every publication must have a contact, so it makes sense for a contact to be in the create publication response.
- Remove `Contact` from `PublicationViewModel`. This means that several frontend components need to make additional get and update calls to get / update contact information.
- The `GetPublication` endpoint now hydrates all parts of the `PublicationViewModel` that it returns, specifically `Permissions` and `IsSuperseded`.
- Add a query parameter for the `GetPublication` endpoint so services can choose whether they want `Permissions` returned.
- Change the frontend's `PublicationContext` to use `Publication` instead of `MyPublication`. This change has meant that some components have had to make an additional call to get other information, like publication's contact information.
- Several frontend interfaces have been renamed to match their associated backend view models.

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/PublicationController.cs [hunk 1]
```diff
@@ -64,10 +64,10 @@ public async Task<ActionResult<List<PublicationSummaryViewModel>>> ListPublicati
 
         [HttpGet("api/publications/{publicationId}")]
         public async Task<ActionResult<PublicationViewModel>> GetPublication(
-            [Required] Guid publicationId)
+            [Required] Guid publicationId, [FromQuery] bool permissions = false)
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/PublicationController.cs:67]: At this point in reading through the PR, it's unclear what the purpose of the "permissions" flag is here.

Presumably it'll decide whether or not to include a user's permissions with the returned object?  If so, I'd be tempted to rename this to `includePermissions` or `includeUserPermissions`.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                oldType: "uniqueidentifier",
+                oldNullable: true);
+
+            migrationBuilder.AddForeignKey(
+                name: "FK_Publications_Contacts_ContactId",
+                table: "Publications",
+                column: "ContactId",
+                principalTable: "Contacts",
+                principalColumn: "Id",
+                onDelete: ReferentialAction.Cascade);
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs:32]: Not 100% sure on this, but I think this is telling EF that if a Contact record is deleted, it'd then try to cascade that and delete its owning Publication too.  Worth checking with other devs?

Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs:32]: Assuming that's so, would suggest changing it to ReferentialAction.Restrict.  This'll mean EF would not allow the deletion of a Contact that is in use by a Publication.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs [hunk 3]
```diff
@@ -369,6 +357,54 @@ await supersededPublications
...
+                    }
+
+                    publication.Contact.ContactName = updatedContact.ContactName;
+                    publication.Contact.ContactTelNo = updatedContact.ContactTelNo;
+                    publication.Contact.TeamName = updatedContact.TeamName;
+                    publication.Contact.TeamEmail = updatedContact.TeamEmail;
+                    await _context.SaveChangesAsync();
+
+                    // Clear cache because Contact is in Content.Services.ViewModels.PublicationViewModel
+                    await _publicationCacheService.UpdatePublication(publication.Slug);
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs:402]: Good spot!

File: src/explore-education-statistics-admin/src/services/publicationService.ts [hunk 4]
```diff
@@ -47,28 +51,36 @@ export interface MyPublication {
...
 export interface Publication {
   id: string;
   title: string;
   summary: string;
   slug: string;
-  contact: PublicationContactDetails;
   theme: IdTitlePair;
   topic: IdTitlePair;
+  supersededById?: string;
+  isSuperseded?: boolean;
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/services/publicationService.ts:82]: Could this just be inferred from the existence of a non-null `supersededById`?  Not a biggie, probably makes it more readable...

File: src/explore-education-statistics-admin/src/services/publicationService.ts [hunk 5]
```diff
@@ -117,8 +138,13 @@ const publicationService = {
     return client.put(`/publications/${publicationId}`, publication);
   },
 
-  getPublication(publicationId: string): Promise<Publication> {
-    return client.get<Publication>(`/publications/${publicationId}`);
+  getPublication(
+    publicationId: string,
+    permissions = false,
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/services/publicationService.ts:143]: Again, I'd suggest making the name of this variable a bit more explanatory - `includePermissions` or summat.

File: src/explore-education-statistics-admin/src/pages/publication/PublicationDetailsPage.tsx [hunk 6]
```diff
@@ -15,39 +16,29 @@ import React from 'react';
...
-      theme,
-      topic,
-      supersedingPublication,
-    };
-  });
-
-  const { theme, topic, supersedingPublication } = value ?? {};
+      return publicationService.getPublication(supersededById);
+    },
+  );
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/publication/PublicationDetailsPage.tsx:37]: Does `supersededById` need to be added to a dependency array here in `useAsyncHandledRetry`?  Not 100% sure if it needs it in the same way that useEffect() needs it, but would suggest just looking to see if there are any warnings about this in the console when on this page.

File: src/explore-education-statistics-admin/src/services/publicationService.ts (hunk 4 above)
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/services/publicationService.ts:82]: Unfortunately not. For `isSuperseded` to be true, the `supersededById` publication must also have a live release.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/PublicationController.cs (hunk 1 above)
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/PublicationController.cs:67]: Disagree somewhat here. We're already using this flag for a bunch of other endpoints, so this would introduce inconsistency at this point. Not sure it's worth the effort to change this.

Personally, don't think it's particularly unclear what the purpose is and it's not super hard to follow it through to figure it out. This said, could be convinced that `includePermissions` is better, but `includeUserPermissions` seems like overkill.

Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/PublicationController.cs]: Discussed this on Slack. I've switched to using `includePermission` and changed it for all endpoints.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs (hunk 2 above)
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs:32]: I haven't tested it, but I think you're right. I've changed it to `Restrict`.

Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs:32]: Ok Ben said that this actually means if a Publication gets deleted, the associated Contact will be deleted. We think this is fine so I've changed this back to Cascade. I will test this to make sure it's right

File: src/explore-education-statistics-admin/src/pages/publication/PublicationDetailsPage.tsx (hunk 6 above)
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/publication/PublicationDetailsPage.tsx:37]: I've added it to the deps - other places we use `useAsyncHandledRetry` do include variables used in side in the deps.

File: src/explore-education-statistics-admin/src/pages/publication/PublicationContactPage.tsx [hunk 7]
```diff
@@ -11,54 +10,29 @@ import SummaryListItem from '@common/components/SummaryListItem';
...
-      contact: {
-        teamName,
-        teamEmail,
-        contactName,
-        contactTelNo,
-      },
-    });
+    await publicationService.updateContact(publicationId, updatedContact);
 
     onReload();
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/PublicationContactPage.tsx:29]: Don't need this anymore as contact isn't on publication. We actually need to update the `contact` and switch back to the read-only mode during the submit by doing something like this:

```ts
const { value: contact, setValue: setContact } = useAsyncHandledRetry(...);

const handleSubmit = async (updatedContact: PublicationContactFormValues) => {
  ...

  const nextContact = await publicationService.updateContact(publicationId, updatedContact);

  setContact(nextContact);
  toggleReadOnly.off();
};
```

It also seems like we're missing a test case to check that the UI switches back to the read-only state upon successful submit as well.

File: src/explore-education-statistics-admin/src/pages/publication/__tests__/PublicationContactPage.test.tsx [hunk 8]
```diff
@@ -244,23 +260,15 @@ describe('PublicationContactPage', () => {
...
-        {
-          ...testPublication,
-          contact: {
-            contactName: 'John Smith',
-            contactTelNo: '0777777777',
-            teamEmail: 'john.smith@test.com',
-            teamName: 'Team Smith',
-          },
-        },
+        testContact,
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/__tests__/PublicationContactPage.test.tsx:265]: Can appreciate you didn't write this test originally, but it might be a good basic check to change one of the contact form's fields to check that the updated value gets passed through correctly.

File: src/explore-education-statistics-admin/src/services/publicationService.ts [hunk 9]
```diff
@@ -117,8 +138,13 @@ const publicationService = {
     return client.put(`/publications/${publicationId}`, publication);
   },
 
-  getPublication(publicationId: string): Promise<Publication> {
-    return client.get<Publication>(`/publications/${publicationId}`);
+  getPublication(
+    publicationId: string,
+    includePermissions = false,
+  ): Promise<Publication> {
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/services/publicationService.ts:144]: To make this a little more flexible so that we don't need to handle optional permissions in the places where we call this, would suggest we add a generic type for this (similar to the `listReleases` method). It should look something like:
```ts
listReleases<TPublication extends Publication = Publication>(...): Promise<TPublication>
```

Then at call sites, we can do this:
```ts
getPublication<PublicationWithPermissions>(publicationId, true); 
```

Where the return type will now be a `PublicationWithPermissions` rather than a `Publication`

File: src/explore-education-statistics-admin/src/pages/publication/contexts/PublicationContext.tsx [hunk 10]
```diff
@@ -1,10 +1,10 @@
-import { MyPublication } from '@admin/services/publicationService';
+import { Publication } from '@admin/services/publicationService';
 import React, { createContext, ReactNode, useContext, useMemo } from 'react';
 
 export interface PublicationContextState {
-  publication: MyPublication;
+  publication: Publication;
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/contexts/PublicationContext.tsx:5]: All of the types in here should be a `PublicationWithPermissions`

File: src/explore-education-statistics-admin/src/pages/publication/PublicationPageContainer.tsx [hunk 11]
```diff
@@ -55,7 +55,7 @@ const PublicationPageContainer = ({
     isLoading: loadingPublication,
     retry: reloadPublication,
   } = useAsyncHandledRetry(() =>
-    publicationService.getMyPublication(publicationId),
+    publicationService.getPublication(publicationId, true),
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/PublicationPageContainer.tsx:58]: As mentioned in `publicationService.ts`, we should add a generic type parameter so that we can call this like:
```ts
publicationService.getPublication<PublicationWithPermissions>(publicationId, true)
```

File: src/explore-education-statistics-admin/src/pages/publication/PublicationDetailsPage.tsx [hunk 12]
```diff
@@ -15,39 +16,30 @@ import React from 'react';
...
-    themeId,
     title,
     summary,
-    topicId,
-  } = publication;
+    permissions,
+    supersededById,
+    theme,
+    topic,
+  } = publication as Publication;
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/PublicationDetailsPage.tsx:26]: We shouldn't need this `as Publication`

File: src/explore-education-statistics-admin/src/pages/publication/PublicationDetailsPage.tsx [hunk 13]
```diff
@@ -70,7 +62,7 @@ const PublicationDetailsPage = () => {
                 : 'This publication is not archived'}
             </SummaryListItem>
           </SummaryList>
-          {permissions.canUpdatePublication && (
+          {permissions?.canUpdatePublication && (
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/PublicationDetailsPage.tsx:65]: As mentioned in `PublicationContact`, we should be working with a `PublicationWithPermissions` so none of the `permissions` should be optional. Same for below this as well.

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/ContactViewModels.cs [hunk 14]
```diff
@@ -1,34 +1,34 @@
...
 
-        public string ContactName { get; set; }
+        public string ContactTelNo { get; set; } = string.Empty;
 
-        public string ContactTelNo { get; set; }
+        public ContactPermissions? Permissions { get; set; }
+
+        public record ContactPermissions
+        {
+            public bool CanUpdatePublication { get; set; }
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/ContactViewModels.cs:20]: Why do we have this on a contact? Shouldn't we have a permission on publication permissions called `CanUpdateContact`?

File: src/explore-education-statistics-admin/src/pages/publication/PublicationMethodologiesPage.tsx [hunk 15]
```diff
@@ -99,7 +99,7 @@ const PublicationMethodologiesPage = () => {
             to an external file that contains methodology details.
           </p>
 
-          {permissions.canCreateMethodologies && (
+          {permissions?.canCreateMethodologies && (
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/PublicationMethodologiesPage.tsx:102]: Again, these shouldn't optional - we should be working with `PublicationWithPermissions` from `PublicationContext`

File: src/explore-education-statistics-admin/src/pages/publication/PublicationReleasesPage.tsx [hunk 16]
```diff
@@ -19,7 +19,7 @@ const PublicationReleasesPage = () => {
 
       <p>View, edit or amend releases contained within this publication.</p>
 
-      {publication?.permissions.canCreateReleases && (
+      {publication?.permissions?.canCreateReleases && (
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/PublicationReleasesPage.tsx:22]: Again, these shouldn't optional - we should be working with `PublicationWithPermissions` from `PublicationContext`

File: src/explore-education-statistics-admin/src/pages/publication/PublicationContactPage.tsx [hunk 17]
```diff
@@ -88,7 +62,7 @@ const PublicationContactPage = () => {
               {contact.contactTelNo}
             </SummaryListItem>
           </SummaryList>
-          {publication.permissions.canUpdatePublication && (
+          {contact.permissions?.canUpdatePublication && (
```
Comment (ntsim) [inline src/explore-education-statistics-admin/src/pages/publication/PublicationContactPage.tsx:65]: As mentioned in another comment, the permission for updating a contact should be on the `publication` itself i.e. we should have something like `publication.permissions.canUpdateContact`.

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs (hunk 2 above)
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs:32]: We've had a discussion on Slack - @markhiveit to test this out to be sure of the correct way round.

File: src/explore-education-statistics-admin/src/services/publicationService.ts (hunk 4 above)
Comment (duncan-at-hiveit) [reply src/explore-education-statistics-admin/src/services/publicationService.ts:82]: I might be tempted to rename `isSuperseded` to something more explicit then, like `isSupersededByLivePublication` or put a code comment next to `isSuperseded` to explain the situation - cos for me personally I'd be confused in a scenario whereby `isSuperseded` is false but `supersededById` is not null :)

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs (hunk 2 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs:32]: Ignore what I said now, I was thinking this FK was being added to the contacts table, but it's a FK on the publications table.

So it does want to be `onDelete: ReferentialAction.Restrict` but we are left with a potential problem of leaving orphaned contacts if publications are deleted.

Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs:32]: Tested this.

`DELETE  FROM Contacts WHERE Id = '35424c89-7ee1-4951-b1e8-f786c69e137c';` also deleted the publication
`DELETE FROM Publications WHERE ContactId = '35424c89-7ee1-4951-b1e8-f786c69e137c';` didn't result in the Contact being deleted.

I'll change the ReferentialAction to Restrict.

Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Migrations/ContentMigrations/20220914155611_EES3695MakePublicationsContactIdNonnullable.cs]: Switched to Restrict and reset the test data locally to reapply the migration. Tried to delete a contact and it wouldn't let me while the associated publication still existed.

File: src/explore-education-statistics-admin/src/services/publicationService.ts (hunk 4 above)
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/services/publicationService.ts:82]: I've added a comment in both the backend and frontend that should make this clearer. With hindsight, I think I would have named it `isSupersededByLivePublication`, as that clears up the ambiguity. But I'm inclined not to rename it now as that means invalidating all publications' publication.json. And since this work is likely to be deployed alongside the read replica work, we probably want as few things to worry about in that deploy as possible.

But still happy to rename it if you'd prefer - or I could create a ticket to rename it at a later date.

Comment (ntsim) [reply src/explore-education-statistics-admin/src/services/publicationService.ts:82]: Just to chime in here, but a more terse suggestion would be `isSupersededByLive`

File: src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/ContactViewModels.cs (hunk 14 above)
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/ContactViewModels.cs:20]: I've just changed the name to `CanUpdateContact` for now - it still uses `userService.CheckCanUpdatePublication`. The same work will also need to be done with external methodology, legacy releases, etc., so I intend to sort out all the permission changes in my current branch. I've left a comment in `PermissionsUtils` to remind me to do this for contacts.

File: src/explore-education-statistics-admin/src/pages/publication/PublicationContactPage.tsx (hunk 17 above)
Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/publication/PublicationContactPage.tsx:65]: I've also changed `getContact` and added a `ContactWithPermissions` interface so I can remove the `?` here.

Comment (markhiveit) [reply src/explore-education-statistics-admin/src/pages/publication/PublicationContactPage.tsx:65]: Turns out I cannot do that because `UpdateContact` just returns a `Contact` and we pass the result of `UpdateContact` to `setContact` that was expecting a `ContactWithPermissions`. I've left the extra code around `getContact` and the new `ContactWithPermissions` interface.

Comment (ntsim) [reply src/explore-education-statistics-admin/src/pages/publication/PublicationContactPage.tsx:65]: Would really just move the contact permission onto the publication as part of this PR to save yourself the trouble of having to do it later!

File: src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/ContactViewModels.cs (hunk 14 above)
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/ContactViewModels.cs:20]: Would suggest just doing it as part of this PR. Like I'm not sure what time we're saving here by not doing it here?

Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/ViewModels/ContactViewModels.cs]: Discussed this with Nick on Slack - I'll make the changes on this PR.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Util/PermissionsUtils.cs [hunk 18]
```diff
@@ -37,6 +38,7 @@ public static async Task<ReleasePermissions> GetReleasePermissions(IUserService
             CanCreateMethodologies = await userService.CheckCanCreateMethodologyForPublication(publication).IsRight(),
             CanManageExternalMethodology =
                 await userService.CheckCanManageExternalMethodologyForPublication(publication).IsRight(),
+            CanUpdateContact = await userService.CheckCanUpdatePublication(publication).IsRight(),
```
Comment (ntsim) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Util/PermissionsUtils.cs:41]: We already check `userService.CheckCanUpdatePublication` above, so we should consolidate this into one call to avoid an additional database query.

**Review by ntsim — CHANGES_REQUESTED** [review]

**Review by ntsim — APPROVED** [review]
